dashboard/lib/email_reminder.rb

```diff
@@ -0,0 +1,136 @@
+require 'stringio'
+require 'cdo/aws/metrics'
+require 'cdo/aws/s3'
+require 'cdo/chat_client'
+
+class EmailReminder
+  include Rails.application.routes.url_helpers
+  attr_reader :max_reminder_age, :min_reminder_age, :max_reminders, :log
+
+  def initialize(options = {})
+    # Oldest requests we want to send reminders for. Defaults to 7 days ago.
+    @max_reminder_age = options[:max_reminder_age] || 7.days.ago
+    raise ArgumentError.new('max_reminder_age must be Time') unless @max_reminder_age.is_a? Time
+
+    # Newest requests we want to send reminders for. Defaults to 3 days ago.
+    @min_reminder_age = options[:min_reminder_age] || 3.days.ago
+    raise ArgumentError.new('min_reminder_age must be Time') unless @min_reminder_age.is_a? Time
+    raise ArgumentError.new('max_reminder_age must come before min_reminder_age') unless @max_reminder_age < @min_reminder_age
+
+    # Max reminders we want to send for a single permission request. Defaults to 1.
+    @max_reminders = options[:max_reminders] || 1
+    raise ArgumentError.new('max_reminders must be Integer') unless @max_reminders.is_a? Integer
+
+    reset
+  end
+
+  # Find permission requests for users who haven't been granted permission yet.
+  # Only find requests that are between max_reminder_age and min_reminder_age old.
+  # Only find requests that have sent fewer than max_reminders reminders.
+  # Return the IDs of the requests.
+  def find_requests_needing_reminder
+    ParentalPermissionRequest.joins(:user).
+      select(:id).
+      where('parental_permission_requests.created_at BETWEEN ? AND ?', @max_reminder_age, @min_reminder_age).
+      where('parental_permission_requests.reminders_sent < ?', @max_reminders).
+      where("JSON_EXTRACT(users.properties, '$.child_account_compliance_state') != ?", User::ChildAccountCompliance::PERMISSION_GRANTED)
+  end
+
+  # Send a reminder for a given ParentalPermissionRequest ID.
+  # Increment the reminders_sent count.
+  def send_permission_reminder_email(request_id)
+    request = ParentalPermissionRequest.find(request_id)
+    permission_url = url_for(controller: :policy_compliance, action: :child_account_consent, host: CDO.studio_url('', CDO.default_scheme), token: request.uuid)
+    ParentMailer.parent_permission_reminder(request.parent_email, permission_url).deliver_now
+    request.reminders_sent += 1
+    request.save!
+  end
+
+  # Send emails for all requests that need reminders.
+  def send_all_reminder_emails
+    find_requests_needing_reminder.each do |request|
```

Comment:
I would recommend using `#find_each` in the case when there are a lot of records
```suggestion
    find_requests_needing_reminder.find_each do |request|
```